File: ArchitectureModel/src/main/java/com/kailin/architecture_model/util/UUIDUtil.java
package com.kailin.architecture_model.util;

import android.content.Context;
import android.provider.Settings;
import android.text.TextUtils;

import java.util.UUID;
import java.util.concurrent.atomic.AtomicReference;

public final class UUIDUtil {

    private static final String PERFERENCES_KEY_UUID = "UUID";

    private static final AtomicReference<UUIDUtil> reference = new AtomicReference<>();

    public static UUIDUtil getInstance(Context context) {
        while (true) {
            UUIDUtil instance = reference.get();
            if (instance != null)
                return instance;

            instance = new UUIDUtil(context);
            if (reference.compareAndSet(null, instance))
                return instance;
        }
    }

    private final Context context;
    private final SharedPreferencesUtil sharedPreferencesUtil;

    private UUIDUtil(Context context) {
        this.context = context.getApplicationContext();
        this.sharedPreferencesUtil = SharedPreferencesUtil.getInstance(context);
    }

    public String getUUID() {
        String uuid = sharedPreferencesUtil.getString(PERFERENCES_KEY_UUID);
        if (TextUtils.isEmpty(uuid)) {
            String androidId = Settings.Secure.getString(context.getContentResolver(), Settings.Secure.ANDROID_ID);
            uuid = UUID.nameUUIDFromBytes(androidId.getBytes()).toString();
            if (TextUtils.isEmpty(uuid)) {
                uuid = UUID.randomUUID().toString();
            }
            sharedPreferencesUtil.putString(PERFERENCES_KEY_UUID, uuid);
        }
        return uuid;
    }
}

File: app/src/main/java/com/kailin/utillibrary/UUIDUtil.kt
package com.kailin.utillibrary

import android.content.Context
import android.provider.Settings
import com.kailin.utillibrary.security.SecurityPreferencesUtil
import java.util.concurrent.atomic.AtomicReference

class UUIDUtil private constructor(private val context: Context) {

    val UUID: String
        get() {
            val spUtil = SecurityPreferencesUtil.getInstance(context)
            var uuid = spUtil.getString(UUID_UTIL_SP_UUID_VALUE)
            if (uuid.isEmpty()) {
                val androidId = Settings.Secure.getString(context.contentResolver, Settings.Secure.ANDROID_ID)
                uuid = java.util.UUID.nameUUIDFromBytes(androidId.toByteArray()).toString()
                spUtil.putString(UUID_UTIL_SP_UUID_VALUE, uuid)
            }
            return uuid
        }

    companion object {
        private const val UUID_UTIL_SP_UUID_VALUE = "UUID_UTIL_SP_UUID_VALUE"
        private val reference = AtomicReference<UUIDUtil>()

        fun getInstance(context: Context): UUIDUtil {
            while (true) {
                var uuidUtil: UUIDUtil? = reference.get()
                if (uuidUtil != null)
                    return uuidUtil

                uuidUtil = UUIDUtil(context)
                if (reference.compareAndSet(null, uuidUtil))
                    return uuidUtil
            }
        }
    }
}